### ISSUE TITLE
Luxid fails if DNS returns an IPv6 address that does not reverse resolve

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#1115]
Originally reported of Google Code with ID 1059.

```
<b>What steps will reproduce the problem?</b>
1. Set up DNS so that the master node has both, an ipv4 and an ipv6 address assigned.
2. Configure the host in a way, that it only has the ipv4 address.

For example by disabling ipv6:
sysctl -w net.ipv6.conf.eth0.disable_ipv6=1

3. Create a one-node cluster on that node.
4. Run gnt-cluster info

<b>What is the expected output? What do you see instead?</b>

I see:

gnt-cluster info
Cannot communicate with the config daemon.
Is the process running and listening for connections?

While I would have expected, well, the cluster info.

It turns out, that luxid does not start. Here's an error from the commands.log:

2015-04-09 16:46:45,321: gnt-cluster info pid=32329 ERROR Error during command processing
Traceback (most recent call last):
  File "/usr/local/share/ganeti/2.10/ganeti/cli.py", line 2597, in GenericMain
    result = func(options, args)
  File "/usr/local/share/ganeti/2.10/ganeti/client/gnt_cluster.py", line 494, in ShowClusterConfig
    cl = GetClient(query=True)
  File "/usr/local/share/ganeti/2.10/ganeti/cli.py", line 2416, in GetClient
    client = luxi.Client(address=address)
  File "/usr/local/share/ganeti/2.10/ganeti/luxi.py", line 410, in __init__
    self._InitTransport()
  File "/usr/local/share/ganeti/2.10/ganeti/luxi.py", line 418, in _InitTransport
    timeouts=self.timeouts)
  File "/usr/local/share/ganeti/2.10/ganeti/luxi.py", line 177, in __init__
    args=(self.socket, address, self._ctimeout))
  File "/usr/local/share/ganeti/2.10/ganeti/utils/retry.py", line 173, in Retry
    return fn(*args)
  File "/usr/local/share/ganeti/2.10/ganeti/luxi.py", line 198, in _Connect
    raise NoMasterError(address)
NoMasterError: /var/run/ganeti/socket/ganeti-query
~                                                        

This is only vaguely researched, but maybe it helps:
The culprit seems to be getFQDN from Daemon.hs which, when ipv6 for DNS is enabled, looks at that address, compares it to the ipv4, and decides 'that is not me, I am not master."

A possible solution would be to do a DNS look up and pick the address that is matching to the address family in the cluster config.

A workaround for now is to put the host and ipv4 address in the /etc/hosts of the node.

```
Originally added on 2015-04-09 16:56:41 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Further investigations show that while Network.Socket.getNameInfo properly
looks up PTR for ipv4 addresses, it seems to fail to do so for ipv6 addresses.
```
Originally added on 2015-04-10 09:02:41 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
...and getNameInfo is just a thin wrapper around getnameinfo(3). So
maybe the PTR record is missing after all.
```
Originally added on 2015-04-10 09:16:02 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Just confirmed that the IPv6 PTR record is missing for the machine where
the error occured.
```
Originally added on 2015-04-10 09:28:32 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2015-04-10 09:29:01 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Will add the ip-family of the cluster as preferred family
when resolving the name.
```
Originally added on 2015-04-10 09:30:28 +0000 UTC.

Changed State: Started

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2015-04-10 09:30:51 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
commit b084f5519992ffed480ab24fe8194fd830a3b0e7
Author: Klaus Aehlig <aehlig@google.com>
Date:   Fri Apr 10 12:15:24 2015 +0200

    Make getFQDN prefer cluster protocol family
    
    In getFQDN, if the primary IP family of the cluster is known,
    provide it as a hint.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

commit baecc91fae408187017fb71ebc96a27f1d8a2f66
Author: Klaus Aehlig <aehlig@google.com>
Date:   Fri Apr 10 12:05:50 2015 +0200

    Add version of getFQDN accepting preferences
    
    Still the best way to get the "canonical name" of a host
    is to look up its host name and reverse look up the IP
    address obtained. Obviously, that the result depends
    on the protocol family used. So accept hints on which
    family to prefer.
    
    Note that getFQDN is exported on higher branches, so we
    keep its type stable to avoid semantic merge conflicts.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

commit cf1bf3add69fc6ba0309ba4bdc3c08af7f21a52f
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Jun 2 17:31:10 2014 +0200

    Make getFQDN honor vcluster
    
    Make the official (exported) getFQDN function honor the vcluster
    setup. In this way, also Haskell daemons can obtain their name
    correctly in a vcluster. In particular, voting will work.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>
    
    Cherry-picked-form: c29501f6
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

```
Originally added on 2015-04-10 11:01:38 +0000 UTC.

Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Released as part of 2.14.0
```
Originally added on 2015-06-02 10:40:20 +0000 UTC.

Changed State: Released
